### PR TITLE
Inherit setter values from parent package

### PR DESCRIPTION
**Pull request by phanimarupaka**
@mortent @morgante 

This PR is to organize auto-setters and add new feature to auto-setters to inherit the values of parent package setters when the package is being fetched during get and update operations. The setters with isSet not true in setter definitions of fetched/updated packages are only auto-set

### CODE REVIEW

File: internal/cmdget/cmdget.go [hunk 1]
```diff
@@ -87,8 +88,9 @@ func (r *Runner) runE(c *cobra.Command, args []string) error {
 	}
 
 	if r.AutoSet {
-		if err := setters.PerformAutoSetters(r.Get.Destination); err != nil {
-			return err
+		if err := setters.PerformAutoSetters(r.Get.Destination, filepath.Dir(r.Get.Destination)); err != nil {
+			// do nothing as auto-setters are best effort and should not throw an error if failed
```
Comment (mortent) [inline internal/cmdget/cmdget.go:92]: We shouldn't swallow any errors here.

File: internal/util/setters/setters.go [hunk 2]
```diff
@@ -37,14 +37,19 @@ const (
 	GcloudProjectNumber = "gcloud.project.projectNumber"
 )
 
-func PerformAutoSetters(path string) error {
+func PerformAutoSetters(targetPath, parentPath string) error {
```
Comment (mortent) [inline internal/util/setters/setters.go:40]: So here we basically try to set all of them and let the last win? I think it would make more sense if we try them in the correct order and then exit early if we are successful. In this current solution, if the value from env doesn't pass validation, this will fail even if the inherited value (which is the only one we care about) is valid.

File: internal/util/update/resource-merge.go [hunk 3]
```diff
@@ -59,9 +60,15 @@ func (u ResourceMergeUpdater) Update(options UpdateOptions) error {
 	}
 	defer os.RemoveAll(updated.AbsPath())
 
-	err := setters.PerformAutoSetters(updated.AbsPath())
-	if err != nil {
-		return err
+	if options.AutoSet {
```
Comment (mortent) [inline internal/util/update/resource-merge.go:63]: We should probably add a test for each of the scenarios outlined in the design doc.

File: internal/cmdget/cmdget.go (hunk 1 above)
Comment (phanimarupaka) [reply internal/cmdget/cmdget.go]: Done. Printed error messages to stdout for each failure and continued. Fatal errors are bubbled up.

File: internal/util/setters/setters.go (hunk 2 above)
Comment (phanimarupaka) [reply internal/util/setters/setters.go]: Done. Leveraged isSet flag.

File: internal/util/update/resource-merge.go (hunk 3 above)
Comment (phanimarupaka) [reply internal/util/update/resource-merge.go]: Updated the logic. Assuming update logic works correctly and well tested, this change would only set auto-setters for the merged package if isSet flag is not true. 

All test cases related to updated can be found here https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/yaml/merge3/merge3_test.go

File: internal/util/setters/setters.go [hunk 4]
```diff
@@ -133,58 +150,54 @@ func SetV1AutoSetter(name, value, path string) error {
...
 	for _, resourcesPath := range resourcePackagesPaths {
-		fs := &settersutil.FieldSetter{
-			Name:               name,
-			Value:              value,
-			SetBy:              "kpt",
-			OpenAPIPath:        filepath.Join(resourcesPath, kptfile.KptFileName),
-			OpenAPIFileName:    kptfile.KptFileName,
-			ResourcesPath:      resourcesPath,
-			RecurseSubPackages: true,
+		if isSet(name, filepath.Join(resourcesPath, kptfile.KptFileName)) {
```
Comment (mortent) [inline internal/util/setters/setters.go:159]: Is it safe to rely on the `isSet` property here? I'm both thinking if there are situations where to user might have updated the value, but it should still be covered by the autosetter, and if there are situations where an upstream package might already have the `isSet` property set.

File: internal/util/setters/setters.go [hunk 5]
```diff
@@ -262,6 +277,116 @@ func SetEnvAutoSetters(root string) error {
...
+// SetInheritedSetters traverses the absolute parentPath all the way to the root
+// of file system to find a kpt package with Kptfile and auto-fills the setter
+// values in targetPath and all its subpackages
+func (a AutoSet) SetInheritedSetters() error {
+	parentKptfilePath, err := parentDirWithKptfile(a.ParentPath)
+	if parentKptfilePath == "" || err != nil {
+		return err
+	}
+
+	if err := openapi.AddSchemaFromFile(parentKptfilePath); err != nil {
```
Comment (mortent) [inline internal/util/setters/setters.go:289]: Do we need to clean this up?

File: internal/util/setters/setters.go [hunk 6]
```diff
@@ -29,30 +31,45 @@ import (
...
-func PerformAutoSetters(path string) error {
-	// auto-fill setters from environment
-	if err := SetEnvAutoSetters(path); err != nil {
+type AutoSet struct {
+	Writer     io.Writer
+	TargetPath string
+	ParentPath string
+}
+
+// PerformAutoSetters auto-fills the setter values from the local environment
```
Comment (mortent) [inline internal/util/setters/setters.go:50]: Can we explain in some more detail here what is happening? In particular how we handle different types of autosetters and how we handle subpackage.

File: internal/util/setters/setters.go (hunk 4 above)
Comment (phanimarupaka) [reply internal/util/setters/setters.go:159]: `isSet` is only checked for the local package which is being set. So the second point you mentioned is taken care of.

I don't see any cases where we should set the value after the value has been set locally.

File: internal/util/setters/setters.go (hunk 6 above)
Comment (phanimarupaka) [reply internal/util/setters/setters.go:50]: Will do.

File: internal/util/setters/setters.go (hunk 5 above)
Comment (phanimarupaka) [reply internal/util/setters/setters.go:289]: Good point. Will do.

File: internal/util/openapi/openapi.go [hunk 7]
```diff
@@ -83,5 +83,5 @@ func ConfigureOpenAPISchema(openAPISchema []byte) error {
 	// of where we got the Kubernetes openAPI schema.
 	// TODO: Refactor the openapi package in kyaml so we don't need to
 	// know the name of the kustomize asset here.
-	return openapi.AddSchema(kustomizationapi.MustAsset("openapi/kustomizationapi/swagger.json"))
+	return openapi.AddSchema(kustomizationapi.MustAsset("kustomizationapi/swagger.json"))
```
Comment (mortent) [inline internal/util/openapi/openapi.go:86]: Why is this change needed?

File: internal/util/setters/setters.go [hunk 8]
```diff
@@ -29,30 +31,53 @@ import (
...
 
 const (
 	GcloudProject       = "gcloud.core.project"
 	GcloudProjectNumber = "gcloud.project.projectNumber"
 )
 
-func PerformAutoSetters(path string) error {
+type AutoSet struct {
+	Writer     io.Writer
+	TargetPath string
```
Comment (mortent) [inline internal/util/setters/setters.go:46]: Can we clarify the difference between TargetPath and ParentPath?

File: internal/cmdget/cmdget.go [hunk 9]
```diff
@@ -87,7 +88,12 @@ func (r *Runner) runE(c *cobra.Command, args []string) error {
 	}
 
 	if r.AutoSet {
-		if err := setters.PerformAutoSetters(r.Get.Destination); err != nil {
+		a := setters.AutoSet{
+			Writer:     c.OutOrStdout(),
+			TargetPath: r.Get.Destination,
+			ParentPath: filepath.Dir(r.Get.Destination),
```
Comment (mortent) [inline internal/cmdget/cmdget.go:94]: Can we explain this? Is destination here a reference to a file or a folder? And is ParentPath always just the parent folder?

File: internal/util/update/resource-merge.go [hunk 10]
```diff
@@ -106,6 +101,16 @@ func (u ResourceMergeUpdater) Update(options UpdateOptions) error {
 		return err
 	}
 
+	// perform auto-setters after the package is updated
+	a := setters.AutoSet{
```
Comment (mortent) [inline internal/util/update/resource-merge.go:105]: This seems to be part of the ResourceMerge updater strategy. It seems like this behavior should be the same regardless of the update strategy?

File: internal/util/openapi/openapi.go (hunk 7 above)
Comment (phanimarupaka) [reply internal/util/openapi/openapi.go:86]: Due to this change. Asset name was updated.

https://github.com/kubernetes-sigs/kustomize/commit/c1c2725360e704f4b80b7b207c0aa28a65f6eaca#diff-5a746f23b96330133cb42de78e06cfbfaf388ed252512491b44ae2d61e14a548

File: internal/util/setters/setters.go (hunk 8 above)
Comment (phanimarupaka) [reply internal/util/setters/setters.go]: Updated the PR. Renamed TargetPath to PackagePath and added comment. Removed ParentPath parameter as it is always the parent directory of PackagePath.

File: internal/cmdget/cmdget.go (hunk 9 above)
Comment (phanimarupaka) [reply internal/cmdget/cmdget.go]: Commented above.

File: internal/util/update/resource-merge.go (hunk 10 above)
Comment (phanimarupaka) [reply internal/util/update/resource-merge.go]: Agree. Moved it up to update logic.

**Review by mortent — APPROVED** [review]
